src/zenml/zen_server/cloud_utils.py

```diff
@@ -154,12 +155,29 @@ def session(self) -> requests.Session:
             A requests session with the authentication token.
         """
         if self._session is None:
+            # Set up the session's connection pool size to match the server's
+            # thread pool size. This allows the server to cache one connection
+            # per thread, which means we can keep connections open for longer
+            # and avoid the overhead of setting up a new connection for each
+            # request.
+            conn_pool_size = server_config().thread_pool_size
+
             self._session = requests.Session()
             token = self._fetch_auth_token()
             self._session.headers.update({"Authorization": "Bearer " + token})
 
             retries = Retry(total=5, backoff_factor=0.1)
-            self._session.mount("https://", HTTPAdapter(max_retries=retries))
+            self._session.mount(
+                "https://",
+                HTTPAdapter(
+                    max_retries=retries,
+                    # We only use one connection pool to be cached because we
+                    # only communicate with one remote server (the control
+                    # plane)
+                    pool_connections=1,
```

Comment:
For my understanding, is it costly to have more cached hosts? I understand we need only one but want to know what impact this would have.

Reply:
There is no impact in this case. We can set this to 10 or 100, there will only be one connection pool because we only need to connect to one host. This is better covered here, if you want the gruesome details: https://stackoverflow.com/questions/34837026/whats-the-meaning-of-pool-connections-in-requests-adapters-httpadapter

Reply:
Really well explained answer; had this in bookmarks for a while ;) thanks!